bench/nbody.cc

```diff
@@ -0,0 +1,266 @@
+    typename Backend::Float_v p1posy,
+    typename Backend::Float_v p1posz,
+    typename Backend::Float_v& p1velx,
+    typename Backend::Float_v& p1vely,
+    typename Backend::Float_v& p1velz,
+    typename Backend::Float_v p2posx,
+    typename Backend::Float_v p2posy,
+    typename Backend::Float_v p2posz,
+    typename Backend::Float_v p2mass)
+{
+    const typename Backend::Float_v xdistance = p1posx - p2posx;
+    const typename Backend::Float_v ydistance = p1posy - p2posy;
+    const typename Backend::Float_v zdistance = p1posz - p2posz;
+    const typename Backend::Float_v xdistanceSqr = xdistance * xdistance;
+    const typename Backend::Float_v ydistanceSqr = ydistance * ydistance;
+    const typename Backend::Float_v zdistanceSqr = zdistance * zdistance;
+    const typename Backend::Float_v distSqr = eps2 + xdistanceSqr + ydistanceSqr + zdistanceSqr;
+    const typename Backend::Float_v distSixth = distSqr * distSqr * distSqr;
+    const typename Backend::Float_v invDistCube = 1.0f / math::Sqrt(distSixth);
+    const typename Backend::Float_v sts = p2mass * invDistCube * timeStep;
+    p1velx += xdistanceSqr * sts;
+    p1vely += ydistanceSqr * sts;
+    p1velz += zdistanceSqr * sts;
+}
+
+template <class Backend>
+VECCORE_FORCE_NOINLINE
+void TestNBodySIMD(float* __restrict__ posx, float* __restrict__ posy, float* __restrict__ posz,
+                   float* __restrict__ velx, float* __restrict__ vely, float* __restrict__ velz,
+                   float* __restrict__ mass, size_t kN, const char* name)
+{
+    using Float_v = typename Backend::Float_v;
+    using Int32_v = typename Backend::Int32_v;
+
+    Timer<time_unit> timer;
+    unsigned long long t[kNruns];
+    for (size_t n = 0; n < kNruns; n++) {
+        timer.Start();
+
+        for (std::size_t i = 0; i < kN; i++) {
+            for (std::size_t j = 0; j < kN; j += VectorSize<Float_v>()) {
+                pPInteractionSIMD<Backend>((Float_v&)(posx[j]), (Float_v&)(posy[j]), (Float_v&)(posz[j]),
+                                           (Float_v&)(velx[j]), (Float_v&)(vely[j]), (Float_v&)(velz[j]),
+                                           Float_v{posx[i]}, Float_v{posy[i]}, Float_v{posz[i]}, Float_v{mass[i]});
+            }
+        }
+        for (std::size_t i = 0; i < kN; i += VectorSize<Float_v>()) {
+             (Float_v &)(posx[i]) += (const Float_v &)(velx[i]) * timeStep;
+             (Float_v &)(posy[i]) += (const Float_v &)(vely[i]) * timeStep;
+             (Float_v &)(posz[i]) += (const Float_v &)(velz[i]) * timeStep;
+        }
+
+        t[n] = timer.Elapsed();
+    }
+
+    const unsigned long long mean = std::accumulate(std::begin(t), std::end(t), 0ull) / (kN * kNruns);
+    printf("%20s %6llu\n", name, mean);
+}
+
+int main()
+{
+    auto alloc = [] {
+        return std::unique_ptr<float[], decltype(&AlignedFree)>{
+            static_cast<float*>(AlignedAlloc(VECCORE_SIMD_ALIGN, kN * sizeof(float))), &AlignedFree
+        };
+    };
```

Comment:
A bit overkill, but pretty nice :+1:

Reply:
I just could not stand the look of an explicit call to a free function. But that `unique_ptr` is definitely ugly!